### PR TITLE
#409:Make bigtable::TableAdmin CopyAssignable and CopyConstructible

### PR DESCRIPTION
**Pull request by hemant-qlogic**

### CODE REVIEW

File: bigtable/client/internal/table_admin.h [hunk 1]
```diff
@@ -68,6 +68,35 @@ class TableAdmin {
         rpc_backoff_policy_(backoff_policy.clone()),
         metadata_update_policy_(instance_name(), MetadataParamTypes::PARENT) {}
 
+  /**
+   * Create a new TableAdmin using explicit copy constructor implementation.
+   */
+  TableAdmin(const TableAdmin& table_admin)
```
Comment (coryan) [inline bigtable/client/internal/table_admin.h]: `TableAdmin const&`

File: bigtable/client/internal/table_admin.h [hunk 2]
```diff
@@ -68,6 +68,35 @@ class TableAdmin {
...
+      : client_(std::move(table_admin.client_)),
```
Comment (coryan) [inline bigtable/client/internal/table_admin.h]: There is no effect when you move from a `const` object, which these are, please remove.

File: bigtable/client/internal/table_admin.h [hunk 3]
```diff
@@ -68,6 +68,35 @@ class TableAdmin {
...
+        instance_id_(std::move(table_admin.instance_id_)),
+        instance_name_(table_admin.InstanceName()),
+        rpc_retry_policy_(table_admin.rpc_retry_policy_.get()->clone()),
```
Comment (coryan) [inline bigtable/client/internal/table_admin.h]: Alternatively, you could replace the `std::unique_ptr<>` objects with `std::shared_ptr<>` objects and the compiler will generate all the right copy and move operators for you.

File: bigtable/client/internal/table_admin_test.cc [hunk 4]
```diff
@@ -288,6 +288,72 @@ TEST_F(TableAdminTest, CreateTableFailure) {
   EXPECT_FALSE(status.ok());
 }
 
+/**
+ * @test Verify that Copy Constructor and assignment operator
+ * copies all properties.
+ */
+TEST_F(TableAdminTest, CopyConstructibleAssignableTest) {
+  using namespace ::testing;
```
Comment (coryan) [inline bigtable/client/internal/table_admin_test.cc]: Please add something like this:

```C++
static_assert(std::is_copy_assignable<bigtable::noex::TableAdmin, bigtable::noex::TableAdmin>::value,
  "bigtable::noex::TableAdmin must be assignable");
```

File: bigtable/client/table_admin.h [hunk 5]
```diff
@@ -58,6 +58,23 @@ class TableAdmin {
       : impl_(std::move(client), std::move(instance_id),
               std::move(retry_policy), std::move(backoff_policy)) {}
 
+  /**
```
Comment (coryan) [inline bigtable/client/table_admin.h]: The default implementation should work for these, please remove or replace with:

```C++
TableAdmin(TableAdmin const& rhs) = default;
TableAdmin& operator=(TableAdmin const& rhs) = default;
```

File: bigtable/client/table_admin_test.cc [hunk 6]
```diff
@@ -306,6 +306,74 @@ TEST_F(TableAdminTest, CreateTableFailure) {
 #endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 }
 
+/**
+ * @test Verify that Copy Constructor and assignment operator
+ * copies all properties.
+ */
+TEST_F(TableAdminTest, CopyConstructibleAssignableTest) {
+  using namespace ::testing;
```
Comment (coryan) [inline bigtable/client/table_admin_test.cc]: Same thing, add static assertions here on in `table_admin.cc`.

File: bigtable/client/internal/table_admin_test.cc (hunk 4 above)
Comment (coryan) [reply bigtable/client/internal/table_admin_test.cc]: Or you can add the static assert to the `internal/table_admin.cc` file.

File: bigtable/client/internal/table_admin.h [hunk 7]
```diff
@@ -68,6 +68,35 @@ class TableAdmin {
         rpc_backoff_policy_(backoff_policy.clone()),
         metadata_update_policy_(instance_name(), MetadataParamTypes::PARENT) {}
 
+  /**
```
Comment (coryan) [inline bigtable/client/internal/table_admin.h]: You can remove this, the compiler will provide a good implementation.

File: bigtable/client/internal/table_admin.h [hunk 8]
```diff
@@ -68,6 +68,35 @@ class TableAdmin {
...
+  TableAdmin(TableAdmin const& table_admin)
+      : client_(table_admin.client_),
+        instance_id_(table_admin.instance_id_),
+        instance_name_(table_admin.InstanceName()),
+        rpc_retry_policy_(table_admin.rpc_retry_policy_.get()->clone()),
+        rpc_backoff_policy_(table_admin.rpc_backoff_policy_.get()->clone()),
+        metadata_update_policy_(table_admin.instance_name(),
+                                MetadataParamTypes::PARENT) {}
+
+  /**
```
Comment (coryan) [inline bigtable/client/internal/table_admin.h]: Same here.

File: bigtable/client/table_admin.h (hunk 5 above)
Comment (coryan) [reply bigtable/client/table_admin.h]: You can remove the comments, it is pretty obvious what is going on.

**Comment by hemant-qlogic** [comment]
PTAL

File: bigtable/client/table_admin_test.cc (hunk 6 above)
Comment (coryan) [reply bigtable/client/table_admin_test.cc]: I do not see the static asserts for `bigtable::TableAdmin` (without `noex`) did I miss them?

File: bigtable/client/table_admin.h [hunk 9]
```diff
@@ -58,6 +58,10 @@ class TableAdmin {
       : impl_(std::move(client), std::move(instance_id),
               std::move(retry_policy), std::move(backoff_policy)) {}
 
+  TableAdmin(TableAdmin const& table_admin) = default;
+
```
Comment (coryan) [inline bigtable/client/table_admin.h]: Remove the blank line if you can.

File: bigtable/client/internal/table_admin.cc [hunk 10]
```diff
@@ -21,6 +21,11 @@ namespace bigtable {
 inline namespace BIGTABLE_CLIENT_NS {
 namespace noex {
 
+static_assert(std::is_copy_constructible<bigtable::noex::TableAdmin>::value,
+              "bigtable::noex::TableAdmin must be constructible");
+static_assert(std::is_copy_assignable<bigtable::noex::TableAdmin>::value,
```
Comment (hemant-qlogic) [inline bigtable/client/internal/table_admin.cc]: Here is the static_assert added

File: bigtable/client/table_admin.h (hunk 9 above)
Comment (hemant-qlogic) [reply bigtable/client/table_admin.h]: static_assert is added in file "bigtable/client/internal/table_admin.cc" in last commit

**Comment by hemant-qlogic** [comment]
PTAL

File: bigtable/client/internal/table_admin.cc (hunk 10 above)
Comment (coryan) [reply bigtable/client/internal/table_admin.cc]: Right, but that is for `bigtable::noex::TableAdmin`, I was asking for `bigtable::TableAdmin` without the `noex`.

Comment (hemant-qlogic) [reply bigtable/client/internal/table_admin.cc]: Sorry I misunderstood that. Now change it and commit.

**Pull request merged**